### PR TITLE
Feature rpc server

### PR DESCRIPTION
**Pull request by alloncm**
can be CRed only after the netq rpc connection has been merged

### CODE REVIEW

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+using System.Threading.Tasks;
+
+namespace Blizzard.IO.RabbitMQ.Rpc
+{
+    public class NetqRabbitRpcServer : IRpcServer
+    {
+        private readonly INetqRpcRabbitConnection<Func<Type,object>> _netqRpcRabbitConnection;
+        private readonly ILogger _logger;
+
+        private IDisposable _respondHandler;
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs]:
```suggestion
        private readonly INetqRpcRabbitConnection<Func<Type,object>> _netqRpcRabbitConnection;
        private IDisposable _respondHandler;
        private readonly ILogger _logger;
```

I looks more good like that and have some sort of logic, they ordered like the order you use them. And i think that the logger should always be at the end

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        public NetqRabbitRpcServer(INetqRpcRabbitConnection<Func<Type,object>> netqRpcRabbitConnection, ILoggerFactory loggerFactory)
+        {
+            _netqRpcRabbitConnection = netqRpcRabbitConnection;
+            _logger = loggerFactory.CreateLogger(nameof(NetqRabbitRpcServer));
+            _respondHandler = null;
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs]:
```suggestion
            _netqRpcRabbitConnection = netqRpcRabbitConnection;
            _respondHandler = null;
            _logger = loggerFactory.CreateLogger(nameof(NetqRabbitRpcServer));
```

Same as above , initialize the by the same order. Again i think that the logger looks more good in the end

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            }
+            else
+            {
+                _logger.LogWarning("Could not stop responding cause this server hasnt started yet or has already been stopped");
+            }
+        }
+
+        private void VerifyServerHasntStartedYet()
+        {
+            if (_respondHandler != null)
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs]: I don't like that you use the _responseHandler to determinate if the server started - it's not indicative enough. Add _isStarted

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                throw new InvalidOperationException($"Cannot start responding rpc type is invalid, rpc type: {_netqRpcRabbitConnection.RpcMessageType}");
+            }
+
+            _logger.LogInformation($"Started responding async on bus: {_netqRpcRabbitConnection.Bus}, of type {_netqRpcRabbitConnection.RpcMessageType}");
+        }
+
+        public void Stop()
+        {
+            if (_respondHandler != null)
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs]: I don't like that you use the _responseHandler to determinate if the server started - it's not indicative enough. Add _isStarted

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs (hunk 1 above)
Comment (ybs007) [inline Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs:12]: Maybe i missed something but i didn't understand what you do with the responseHandler?

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+using EasyNetQ;
+using Microsoft.Extensions.Logging;
+using Moq;
+using NUnit.Framework;
+using System;
+using System.Threading.Tasks;
+
+namespace Blizzard.IO.RabbitMQ.Tests.Rpc
+{
+    class RequestStub
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: Create a stub folder and move the two classes there

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    {
+        public int Value { get; set; }
+    }
+
+    class RespondStub
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: Create a stub folder and move the two classes there

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    {
+        public int Value { get; set; }
+    }
+
+    public class NetqRabbitRpcServerTests
+    {
+        private NetqRabbitRpcServer _server;
+        private Mock<INetqRpcRabbitConnection<Func<Type, object>>> _netqRpcRabbitConnectionMock;
+        private Mock<IBus> _busMock;
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]:
```suggestion
        private Mock<INetqRpcRabbitConnection<Func<Type, object>>> _netqRpcRabbitConnectionMock;
        private Mock<IBus> _busMock;
        
        private NetqRabbitRpcServer _server;
```

It's more readable like this

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        [SetUp]
+        public void Setup()
+        {
+            _busMock = new Mock<IBus>();
+            _netqRpcRabbitConnectionMock = new Mock<INetqRpcRabbitConnection<Func<Type, object>>>();
+
+            _netqRpcRabbitConnectionMock.Setup(connection => connection.Bus).Returns(_busMock.Object);
+            Mock<ILogger> loggerMock = new Mock<ILogger>();
+            Mock<ILoggerFactory> lfMock = new Mock<ILoggerFactory>();
+            lfMock.Setup(lf => lf.CreateLogger(It.IsAny<string>())).Returns(loggerMock.Object);
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: Arrange all the mock initialization and then all the setups

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+            //Act
+            _server.Stop();
+
+            //Assert
+            handlerMock.Verify(handler => handler.Dispose(), Times.Once);
+        }
+
+        [Test]
+        public void Stop_OnCallingCallingRespondAsyncOrRespond_ShouldNotDisposeTheHandler()
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: On callingCalling?

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+using Moq;
+using NUnit.Framework;
+using System;
+using System.Threading.Tasks;
+
+namespace Blizzard.IO.RabbitMQ.Tests.Rpc
+{
+    class RequestStub
+    {
+        public int Value { get; set; }
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: You never use the value in both of the stubs why do you need it?

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+            lfMock.Setup(lf => lf.CreateLogger(It.IsAny<string>())).Returns(loggerMock.Object);
+
+            _server = new NetqRabbitRpcServer(_netqRpcRabbitConnectionMock.Object, lfMock.Object);
+        }
+
+        [Test]
+        public void Respond_OnValidCallbackAndConcreteRpcType_ShouldStartRespond()
+        {
+            //Arrange
+            Func<RequestStub, RespondStub> callback = request => new RespondStub();
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: You can make this a local function. I think it looks better, what do you think? 
```suggestion
            RespondStub Callback(RequestStub request) => new RespondStub();
```

And the same for all the stub usage

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+            _netqRpcRabbitConnectionMock.Setup(connection => connection.RpcMessageType).Returns(RpcMessageType.Concrete);
+
+            //Act
+            _server.Respond<RequestStub, RespondStub>(callback);
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: The response explicit types in all the test are unnecessary

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+
+            //Assert
+            _busMock.Verify(bus => bus.Respond<RequestStub, RespondStub>(It.IsAny<Func<RequestStub, RespondStub>>()), Times.Once);
+        }
+
+        [Test]
+        public void Respond_OnInValidRpcType_ShouldThrowInvalidOpertionException()
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]:
```suggestion
        public void Respond_OnInValidRpcType_ShouldThrowInvalidOperationException()
```
Typo in operation

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+            //Arrange
+            Func<RequestStub, RespondStub> callback = request => new RespondStub();
+            _netqRpcRabbitConnectionMock.Setup(connection => connection.RpcMessageType).Returns((RpcMessageType)10);
+
+            //Act & Assert
+            Assert.Throws<InvalidOperationException>(() => _server.Respond<RequestStub, RespondStub>(callback));
+        }
+
+        [Test]
+        public void Respond_OnRespondCalledMoreThanOneTime_ShouldThrowInvalidOpertionException()
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]:
```suggestion
        public void Respond_OnRespondCalledMoreThanOneTime_ShouldThrowInvalidOperationException()
```

typo in operation

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+
+            //Act
+            _server.RespondAsync<RequestStub, RespondStub>(callback);
+
+            //Assert
+            _busMock.Verify(bus => bus.RespondAsync<RequestStub, RespondStub>(It.IsAny<Func<RequestStub, Task<RespondStub>>>()), Times.Once);
+        }
+
+        [Test]
+        public void RespondAsync_OnInValidRpcType_ShouldThrowInvalidOpertionException()
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]:
```suggestion
        public void RespondAsync_OnInValidRpcType_ShouldThrowInvalidOperationException()
```

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+            //Arrange
+            Func<RequestStub, Task<RespondStub>> callback = request => Task.Factory.StartNew(() => new RespondStub());
+            _netqRpcRabbitConnectionMock.Setup(connection => connection.RpcMessageType).Returns((RpcMessageType)10);
+
+            //Act & Assert
+            Assert.Throws<InvalidOperationException>(() => _server.RespondAsync<RequestStub, RespondStub>(callback));
+        }
+
+        [Test]
+        public void RespondAsync_OnRespondCalledMoreThanOneTime_ShouldThrowInvalidOpertionException()
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]:
```suggestion
        public void RespondAsync_OnRespondCalledMoreThanOneTime_ShouldThrowInvalidOperationException()
```

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+            _logger = loggerFactory.CreateLogger(nameof(NetqRabbitRpcServer));
+            _respondHandler = null;
+        }
+
+        public void Respond<TRequest, TRespond>(Func<TRequest, TRespond> callback)
+            where TRequest : class
+            where TRespond : class
+        {
+            VerifyServerHasntStartedYet();
+            if (_netqRpcRabbitConnection.RpcMessageType == RpcMessageType.Abstract)
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs:29]: Why were not using a resolver on the MessageType?

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+            _logger.LogInformation($"Started responding on bus: {_netqRpcRabbitConnection.Bus}, of type {_netqRpcRabbitConnection.RpcMessageType}");
+        }
+
+        public void RespondAsync<TRequest, TRespond>(Func<TRequest, Task<TRespond>> callback)
+            where TRequest : class
+            where TRespond : class
+        {
+            VerifyServerHasntStartedYet();
+            if (_netqRpcRabbitConnection.RpcMessageType == RpcMessageType.Abstract)
```
Comment (ybs007) [inline Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs:56]: Why were not using a resolver on the MessageType?

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs (hunk 4 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs:75]: Here Im not checking if the server has started, im verifying that this handler is not `null` so il not get `NullReferenceException`.
I think more field would just make all the synchronization more dificult

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs (hunk 3 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs]: f

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs (hunk 4 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs]: f

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs (hunk 1 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs:12]: I use it to stop the server from responding without disposing the whole bus (and than terminating the other servers and clients that uses the same bus)

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs (hunk 12 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: I think it is more readable like that

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs (hunk 17 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs:29]: cause there are only two types of objects in c# abstract and concrete. it seems to just make the code less readable (also we wouldnt want to inject a dictionary from `RpcMessageType` to `Action` it from outside of the class)

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs (hunk 5 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: f

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs (hunk 6 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: f

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs (hunk 10 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: f

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs (hunk 11 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: did something better.
added two function to the class and im suing them

File: Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs (hunk 18 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ/Rpc/NetqRabbitRpcServer.cs:56]: Like above

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs (hunk 8 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: I did something difrenet cause the logger is not a field so i wanted to separate them

File: Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs (hunk 9 above)
Comment (alloncm) [reply Blizzard.IO.RabbitMQ.Tests/Rpc/NetqRabbitRpcServerTests.cs]: f

**Pull request merged**